File: leetcode/145. Binary Tree Postorder Traversal/Python3/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorder(self, node, path):
        if node:
            self.postorder(node.left, path)
            self.postorder(node.right, path)
            path.append(node.val)

    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        path = []
        self.postorder(root, path)
        return path

File: leetcode/145. Binary Tree Postorder Traversal/solution.cpp
/**
 * 145. Binary Tree Postorder Traversal
 * 
 * The recursive solution is trivial. Can you do it iteratively?
 * 
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;

        if (root == NULL) return res;

        stack<TreeNode*> s1;
        stack<TreeNode*> s2;

        s1.push(root);

        TreeNode *curr;
        while (!s1.empty()) {
            curr = s1.top();
            s1.pop();
            s2.push(curr);
            if (curr->left != NULL) s1.push(curr->left);
            if (curr->right != NULL) s1.push(curr->right);
        }
        
        while (!s2.empty()) {
            curr = s2.top();
            res.push_back(curr->val);
            s2.pop();
        }
        return res;
    }
};
